File: 2 Mathematics/Python/E-Scooter Math.py
from math import sin, atan

m_pi = 3.14159265358979323846264338327950288    # could've also used pi from math library
kmh_mph_ratio = 0.6213712   # used to convert km/h to mph
gravity = 9.81      # m/s^2

motor_rpm = 4000    # rpm at 6S LiPo, can go up to 14S with our ESC
motor_watt = 1125   # watts at 6S LiPo

wheel_dia = 0.2     # meter
scooter_mass = 10   # kg

try:
    print("\n------------- Electric Scooter -------------\n")

    drive_teeth = float(input("  Amount of teeth on the drive sprocket: "))
    driven_teeth = float(input("  Amount of teeth on the driven sprocket: "))

    gear_ratio = driven_teeth/drive_teeth
    print("\n  Gear ratio = %.2f" % gear_ratio)
    print("  Motor RPM = " + str(motor_rpm))

    wheel_rpm = motor_rpm/gear_ratio
    print("  Wheel RPM = %.2f" % wheel_rpm)

    speed_ms = (m_pi * wheel_dia * wheel_rpm) / 60
    print("\n  Theoretical speed (excl. drag) = %.2f m/s" % speed_ms)
    speed_kmh = speed_ms * 3.6
    print("  Theoretical speed (excl. drag) = %.2f km/h" % speed_kmh + " = %.2f mph" % (speed_kmh * kmh_mph_ratio))

    motor_torque = motor_watt / ((2 * m_pi * motor_rpm) / 60)
    print("\n  Motor torque (excl. losses) = %.2f Nm" % motor_torque)
    wheel_torque = motor_torque * gear_ratio
    print("  Wheel torque (excl. losses) = %.2f Nm" % wheel_torque)

    print("\n---------------- Situation ----------------\n")

    hill_percentage = input("  Driving up a hill with percentage (%): ")

    desired_speed = input("  At a constant speed of (km/h): ")
    person_mass = input("  Person's weight (kg): ")
    total_mass = scooter_mass + float(person_mass)

    necessary_watt = total_mass * gravity * sin(atan(float(hill_percentage) / 100)) * (float(desired_speed) / 3.6)
    print("\n  Necessary wattage to go up that hill = %.2f W" % necessary_watt)
    necessary_torque = total_mass * gravity * sin(atan(float(hill_percentage) / 100)) * (wheel_dia / 2)
    print("  Necessary torque to go up that hill = %.2f Nm" % necessary_torque)

    if wheel_torque > necessary_torque:
        print("\n  We've got enough torque to go up that hill!\n")
    else:
        print("\n Not enough torque to go up that hill...")
        print("  Possible solutions:")
        print("\t> Adjust the gear ratio")
        print("\t> Make the wheels diameter smaller")
        print("\t> Buy a bigger, much stronger, motor and battery")
        print("\t> Never try getting up that hill again")
        print("\t> Go to a planet with less gravity")
        print("\t> Lose weight")

except ValueError:
        spacing = "  "
        new_line = "\n"
        error = " ERROR "
        string = "Learn to type a number."
        print(new_line + spacing + (int(((len(string) - len(error))/2))*"-") + error + (int(((len(string) - len(error))/2))*"-"))
        print(new_line + spacing + string)
        print(new_line + spacing + len(string) * "-")

File: 2 Mathematics/C++/E-Scooter/E-Scooter Math.cpp
// Electric-Scooter-Mathematics.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"
#include <iostream>
#include <iomanip>
#include <math.h>

int getInteger()
{
	while (true)
	{
		int x;
		std::cin >> x;
		std::cin.ignore(32767, '\n');

		if (std::cin.fail())
		{
			std::cin.clear();
			std::cin.ignore(32767, '\n');
		}
		else
			return x;
	}
}

double getDouble()
{
	while (true)
	{
		double x;
		std::cin >> x;
		std::cin.ignore(32767, '\n');

		if (std::cin.fail())
		{
			std::cin.clear();
			std::cin.ignore(32767, '\n');
		}
		else
			return x;
	}
}

int main()
{
	std::cout << "\n----------------------- Electric Scooter ----------------------\n" << std::endl;

	std::cout << "  Drive sprocket teeth: ";
	int driveTooth = getInteger();

	std::cout << "  Driven sprocket teeth: ";
	int followTooth = getInteger();

	float gearRatio = static_cast<float>(followTooth) / driveTooth;
	std::cout << std::setprecision(6);
	std::cout << "\n  Gear ratio = " << gearRatio << std::endl;
	std::cout << "  Motor RPM = " << motorRPM << std::endl;

	double wheelRPM = motorRPM / gearRatio;
	std::cout << "  Wheel RPM = " << wheelRPM << std::endl;

	std::cout << std::setprecision(4);

	double speedMS = (M_PI * wheelDia * wheelRPM) / 60;
	std::cout << "\n  Theoretical speed (excl. drag) = " << speedMS << " m/s" << std::endl;

	double speedKMH = speedMS * 3.6;
	std::cout << "  Theoretical speed (excl. drag) = " << speedKMH << " km/h" << "\t= " << speedKMH * kmh_mph_ratio << " mph" << std::endl;

	double motorTorque = motorWatt / ((2 * M_PI * motorRPM) / 60);
	std::cout << "\n  Motor torque (excl. losses) = " << motorTorque << " Nm" << std::endl;

	double wheelTorque = motorTorque * gearRatio;
	std::cout << "  Wheel torque (excl. losses) = " << wheelTorque << " Nm" << std::endl;

	std::cout << "\n-------------------------- Situation --------------------------\n" << std::endl;

	std::cout << std::resetiosflags(std::ios::fixed | std::ios::showpoint);

	{
		std::cout << "  Driving up hill with percentage (%): ";
		double percentage = getDouble();
		std::cout << "  At a constant speed of (km/h): ";
		double velocity = getDouble();
		std::cout << "  Persons weight (kg): ";
		double personWeight = getDouble();
		double mass = personWeight + scooterWeight;

		double necessaryWatt = mass * gravity * sin(atan(percentage / 100)) * (velocity / 3.6);
		std::cout << "\n  Necessary wattage to go up that hill = " << necessaryWatt << " W" << std::endl;

		double necessaryTorque = mass * gravity * sin(atan(percentage / 100)) * (wheelDia / 2);
		std::cout << "  Necessary torque to go up that hill = " << necessaryTorque << " Nm" << std::endl;

		if (wheelTorque > necessaryTorque)
			std::cout << "\n  We've got enough torque to go up that hill!\n" << std::endl;
		else {
			std::cout << "\n  Not enough torque to go up that hill..." << std::endl;
			std::cout << "  Possible solutions:" << std::endl;
			std::cout << "\t> Adjust the gear ratio" << std::endl;
			std::cout << "\t> Make the wheels diameter smaller" << std::endl;
			std::cout << "\t> Buy a bigger, much stronger, motor and battery" << std::endl;
			std::cout << "\t> Never try getting up that hill" << std::endl;
			std::cout << "\t> Go to a planet with less gravity" << std::endl;
			std::cout << "\t> Lose weight\n" << std::endl;
		}

		std::cout << "---------------------------------------------------------------\n" << std::endl;
	}

	system("pause");
	return 0;
}
